src/checkers/inference/solver/backend/z3smt/Z3SmtSolver.java

```diff
@@ -0,0 +1,429 @@
+
+        logger.fine("Now encoding for unsat core dump.");
+        serializeSMTFileContents();
+
+        solvingStart = System.currentTimeMillis();
+        runZ3Solver();
+        solvingEnd = System.currentTimeMillis();
+
+        Statistics.addOrIncrementEntry(
+                "smt_unsat_serialization_time(millisec)", serializationEnd - serializationStart);
+        Statistics.addOrIncrementEntry(
+                "smt_unsat_solving_time(millisec)", solvingEnd - solvingStart);
+
+        List<Constraint> unsatConstraints = new ArrayList<>();
+
+        for (String constraintID : unsatConstraintIDs) {
+            Constraint c = serializedConstraints.get(constraintID);
+            unsatConstraints.add(c);
+        }
+
+        return unsatConstraints;
+    }
+
+    private void serializeSMTFileContents() {
+        // make a fresh solver to contain encodings of the slots
+        solver = ctx.mkOptimize();
+        // make a new buffer to store the serialized smt file contents
+        smtFileContents = new StringBuilder();
+
+        // only enable in non-optimizing mode
+        if (!optimizingMode && getUnsatCore) {
+            smtFileContents.append("(set-option :produce-unsat-cores true)\n");
+        }
+
+        serializationStart = System.currentTimeMillis();
+        encodeAllSlots();
+        encodeAllConstraints();
+        if (optimizingMode) {
+            encodeAllSoftConstraints();
+        }
+        serializationEnd = System.currentTimeMillis();
+
+        logger.fine("Encoding constraints done!");
+
+        smtFileContents.append("(check-sat)\n");
+        if (!optimizingMode && getUnsatCore) {
+            smtFileContents.append("(get-unsat-core)\n");
+        } else {
+            smtFileContents.append("(get-model)\n");
+        }
+        
+        logger.fine("Writing constraints to file: " + constraintsFile);
+
+        writeConstraintsToSMTFile();
+    }
+
+    private void writeConstraintsToSMTFile() {
+        String fileContents = smtFileContents.toString();
+
+        if (!getUnsatCore) {
+            // write the constraints to the file for external solver use
+            FileUtils.writeFile(new File(constraintsFile), fileContents);
+        } else {
+            // write the unsat core constraints to the file for external solver use
+            FileUtils.writeFile(new File(constraintsUnsatCoreFile), fileContents);
+        }
+        // write a copy in append mode to stats file for later bulk analysis
+        FileUtils.appendFile(new File(constraintsStatsFile), fileContents);
+    }
+
+    protected void encodeAllSlots() {
+        // preprocess slots
+        formatTranslator.preAnalyzeSlots(slots);
+        
+        // generate slot constraints
+        for (Slot slot : slots) {
+            if (slot.isVariable()) {
+                BoolExpr wfConstraint = formatTranslator.encodeSlotWellformednessConstraint(slot);
+
+                if (!wfConstraint.simplify().isTrue()) {
```

Comment:
Can the WF constraint be always false? Warn the user about the invalid encoding?
(Isn't it always being true also strange?)

Reply:
Always true is not odd. Investigate always false later.